security/authorization/util/util_test.go

```diff
@@ -0,0 +1,123 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package engine
+
+import (
+	"testing"
+
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/grpc/internal/grpctest"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func (s) TestStringConvert(t *testing.T) {
+	declarations := []*expr.Decl{
+		decls.NewIdent("request.url_path", decls.String, nil),
+		decls.NewIdent("request.host", decls.String, nil),
+		decls.NewIdent("connection.uri_san_peer_certificate", decls.String, nil)}
+	env, err := cel.NewEnv()
+	if err != nil {
+		t.Fatalf("Failed to create the CEL environment")
+	}
+	for _, test := range []struct {
+		desc             string
+		wantEvalOutcome  bool
+		wantParsingError bool
+		wantEvalError    bool
+		expr             string
+		authzArgs        map[string]interface{}
+	}{
+		{
+			desc:            "Test 1 single primitive match",
+			wantEvalOutcome: true,
+			expr:            "request.url_path.startsWith('/pkg.service/test')",
+			authzArgs:       map[string]interface{}{"request.url_path": "/pkg.service/test"},
+		},
+		{
+			desc:            "Test 2 single compare match",
+			wantEvalOutcome: true,
+			expr:            "connection.uri_san_peer_certificate == 'cluster/ns/default/sa/admin'",
+			authzArgs:       map[string]interface{}{"connection.uri_san_peer_certificate": "cluster/ns/default/sa/admin"},
+		},
+		{
+			desc:            "Test 3 single primitive no match",
+			wantEvalOutcome: false,
+			expr:            "request.url_path.startsWith('/pkg.service/test')",
+			authzArgs:       map[string]interface{}{"request.url_path": "/source/pkg.service/test"},
+		},
+		{
+			desc:            "Test 4 primitive and compare match",
+			wantEvalOutcome: true,
+			expr:            "request.url_path == '/pkg.service/test' && connection.uri_san_peer_certificate == 'cluster/ns/default/sa/admin'",
+			authzArgs: map[string]interface{}{"request.url_path": "/pkg.service/test",
+				"connection.uri_san_peer_certificate": "cluster/ns/default/sa/admin"},
+		},
+		{
+			desc:             "Test 5 parse error field not present in environment",
+			wantParsingError: true,
+			expr:             "request.source_path.startsWith('/pkg.service/test')",
+			authzArgs:        map[string]interface{}{"request.url_path": "/pkg.service/test"},
+		},
+		{
+			desc:          "Test 6 eval error argument not included in environment",
+			wantEvalError: true,
+			expr:          "request.url_path.startsWith('/pkg.service/test')",
+			authzArgs:     map[string]interface{}{"request.source_path": "/pkg.service/test"},
+		},
+	} {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+			checked, err := compileStringToCheckedExpr(test.expr, declarations)
+			want, got := test.wantParsingError, err != nil
```

Comment:
You can replace this with a single line:

```
gotChecked, err := compileStringToCheckedExpr(test.expr, declarations)
if (err != nil) != test.wantParsingError {
  t.Fatal(...)
}
```

The `(err != nil) != test.wantErr` where `test.WantErr` is a boolean might be a little hard to read at first, but this is something that you will see all over the place, so you get used to that pattern very soon.

Reply:
Thanks! This looks pretty clean!